xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererOpenGLES.cpp

```diff
@@ -0,0 +1,239 @@
+{
+  if (feature == RENDERFEATURE_STRETCH         ||
+      feature == RENDERFEATURE_ZOOM            ||
+      feature == RENDERFEATURE_VERTICAL_SHIFT  ||
+      feature == RENDERFEATURE_PIXEL_RATIO     ||
+      feature == RENDERFEATURE_POSTPROCESS     ||
+      feature == RENDERFEATURE_ROTATION)
+  {
+    return true;
+  }
+
+  return false;
+}
+
+bool CRPRendererOpenGLES::Supports(ESCALINGMETHOD method) const
+{
+  if (method == VS_SCALINGMETHOD_NEAREST ||
+      method == VS_SCALINGMETHOD_LINEAR)
+  {
+    return true;
+  }
+
+  return false;
+}
+
+void CRPRendererOpenGLES::CreateTexture()
+{
+  glEnable(m_textureTarget);
+
+  glGenTextures(1, &m_textureId);
+
+  glBindTexture(m_textureTarget, m_textureId);
+
+  glTexImage2D(m_textureTarget, 0, GL_LUMINANCE, m_sourceWidth, m_sourceHeight, 0, GL_LUMINANCE, GL_UNSIGNED_BYTE, NULL);
+
+  glTexParameteri(m_textureTarget, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+  glDisable(m_textureTarget);
+}
+
+void CRPRendererOpenGLES::UploadTexture(const uint8_t* textureData, unsigned int width, unsigned int height)
+{
+  glEnable(m_textureTarget);
+
+  const unsigned int bpp = 1;
+  glPixelStorei(GL_UNPACK_ALIGNMENT, bpp);
+
+  const unsigned datatype = GL_UNSIGNED_BYTE;
+
+  glBindTexture(m_textureTarget, m_textureId);
+  glTexSubImage2D(m_textureTarget, 0, 0, 0, width, height, GL_LUMINANCE, datatype, textureData);
+
+  glBindTexture(m_textureTarget, 0);
+
+  glDisable(m_textureTarget);
+}
+
+void CRPRendererOpenGLES::DeleteTexture()
+{
+  if (glIsTexture(m_textureId))
+    glDeleteTextures(1, &m_textureId);
+
+  m_textureId = 0;
+}
+
+void CRPRendererOpenGLES::ClearBackBuffer()
+{
+  glClearColor(m_clearColour, m_clearColour, m_clearColour, 0.0f);
+  glClear(GL_COLOR_BUFFER_BIT);
+  glClearColor(0.0f, 0.0f, 0.0f, 0.0f);
+}
+
+void CRPRendererOpenGLES::Render()
+{
+  // Create texture
+  if (!glIsTexture(m_textureId))
+    CreateTexture();
+
+  // Upload texture
+  UploadTexture(m_texture.data(), m_sourceWidth, m_sourceHeight);
+
+  glDisable(GL_DEPTH_TEST);
+
+  // Render texture
+  glActiveTexture(GL_TEXTURE0);
+  glEnable(m_textureTarget);
+  glBindTexture(m_textureTarget, m_textureId);
+
+  // Try some clamping or wrapping
+  glTexParameteri(m_textureTarget, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+  GLint filter = (m_renderSettings.VideoSettings().GetScalingMethod() == VS_SCALINGMETHOD_NEAREST ? GL_NEAREST : GL_LINEAR);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_MAG_FILTER, filter);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_MIN_FILTER, filter);
+
+  //! @todo Translate OpenGL code to GLES
+  /*
+  glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);
+
+  glBegin(GL_QUADS);
```

Comment:
How do I render an RGB texture with GLES?

Reply:
this is not implemented for GLES

Reply:
Then what do I do? Convert to YUV and use Kodi's GLES YUV code?

Reply:
how have you done it in the past?

Reply:
In the past I used VP's RenderManager. I converted to YUV and handed it off.

Now, with the new renderer, I'm trying to avoid YUV and render the RGB image directly.

Reply:
If the source is RGB, it should be rendered RGB. If renderer does not support this format there are two options: a) implement workaround in renderer, b) disabled RP on this platform
IMO full platform coverage is not a requirement for a feature. Missing parts can be implemented by platform maintainers at a later time.

Reply:
Looks like OpenGL ES supports at least some the formats we want (`RGB555`, `RGB565`): [Source in the spec]https://www.khronos.org/registry/OpenGL/specs/es/2.0/es_full_spec_2.0.pdf#page=127&zoom=auto]
@garbear If I recall correctly libretro cores also need BGRA8 support? If so we can probably use [OES_rgb8_rgba8](https://www.khronos.org/registry/OpenGL/extensions/OES/OES_rgb8_rgba8.txt) and swizzle our way into BGRA8 in the shader.

Reply:
rendering RBG is trivial, it's just not implemented by Kodi's GLES renderer

Reply:
"RGB" isn't very helpful, unless you mean RGB8 since that's the most common.
Anyway, the current renderer simply outputs in **BGRA8**. However, it relies on sws_scale to do the conversion from the possible _input_ formats taken from libretro cores.

My point above was about using those formats as the format _of the backbuffer_, so that we can achieve "zero copy" rendering and avoid sws_scale/software scaling altogether.
I looked into that for the D3D renderer, but these formats are only supported on DXGI 1.2+ devices unfortunately (no D3D 9), so we didn't go for it.